python/ray/serve/_private/deployment_scheduler.py

```diff
@@ -144,12 +144,25 @@ def required_resources(self) -> Resources:
             self.placement_group_bundles is not None
             and self.placement_group_strategy == "STRICT_PACK"
         ):
-            return sum(
+            required = sum(
                 [Resources(bundle) for bundle in self.placement_group_bundles],
                 Resources(),
             )
         else:
-            return Resources(self.actor_resources)
+            required = Resources(self.actor_resources)
+
+        # Using implicit resource (resources that every node
+        # implicitly has and total is 1)
+        # to limit the number of replicas on a single node.
+        if self.max_replicas_per_node is not None:
+            deployment_id = self.replica_id.deployment_id
+            implicit_resource = (
+                f"{ray._raylet.IMPLICIT_RESOURCE_PREFIX}"
+                f"{deployment_id.app_name}:{deployment_id.name}"
+            )
+            required[implicit_resource] = 1.0 / self.max_replicas_per_node
```

Comment:
What is this magic about? Where is the implicit resource generated?

Reply:
This is generated in core, by default all nodes have 1.0 amount of resources that start with a certain prefix. cc: @jjyao